jlm/hls/backend/rvsdg2rhls/decouple-mem-state.cpp

```diff
@@ -0,0 +1,514 @@
+    {
+      // start of theta
+      state_edge = get_mem_state_user(sn->output(0));
+    }
+    else if (
+        TryGetOwnerOp<llvm::MemoryStateSplitOperation>(*state_edge)
+        || TryGetOwnerOp<llvm::LambdaEntryMemoryStateSplitOperation>(*state_edge))
+    {
+      for (size_t i = 0; i < sn->noutputs(); ++i)
+      {
+        auto followed = follow_state_edge(get_mem_state_user(sn->output(i)), mem_ops, modify);
+        JLM_ASSERT(followed);
+        JLM_ASSERT(
+            TryGetOwnerOp<llvm::MemoryStateMergeOperation>(*followed)
+            || TryGetOwnerOp<llvm::LambdaExitMemoryStateMergeOperation>(*followed));
+        state_edge = get_mem_state_user(followed);
+      }
+    }
+    else if (
+        TryGetOwnerOp<llvm::MemoryStateMergeOperation>(*state_edge)
+        || TryGetOwnerOp<llvm::LambdaExitMemoryStateMergeOperation>(*state_edge))
+    {
+      return sn->output(0);
+    }
+    else if (TryGetOwnerOp<state_gate_op>(*state_edge))
+    {
+      mem_ops.push_back(sn);
+      state_edge = get_mem_state_user(sn->output(si->index()));
+    }
+    else if (TryGetOwnerOp<llvm::LoadNonVolatileOperation>(*state_edge))
+    {
+      mem_ops.push_back(sn);
+      state_edge = get_mem_state_user(sn->output(1));
+    }
+    else if (TryGetOwnerOp<llvm::CallOperation>(*state_edge))
+    {
+      mem_ops.push_back(sn);
+      state_edge = get_mem_state_user(sn->output(sn->noutputs() - 1));
+    }
+    else if (TryGetOwnerOp<llvm::StoreNonVolatileOperation>(*state_edge))
+    {
+      mem_ops.push_back(sn);
+      JLM_ASSERT(sn->output(0)->nusers() == 1);
+      state_edge = get_mem_state_user(sn->output(0));
+      // handle case of store that has one edge going off to deq
+      if (TryGetOwnerOp<llvm::MemoryStateSplitOperation>(*state_edge))
+      {
+        // output 0 is the normal edge
+        state_edge =
+            get_mem_state_user(rvsdg::TryGetOwnerNode<rvsdg::SimpleNode>(*state_edge)->output(0));
+      }
+    }
+    else
+    {
+      JLM_UNREACHABLE("whoops");
+    }
+  }
+}
+
+void
+convert_loop_state_to_lcb(rvsdg::input * loop_state_input)
+{
+  JLM_ASSERT(rvsdg::is<rvsdg::StateType>(loop_state_input->type()));
+  JLM_ASSERT(rvsdg::TryGetOwnerNode<loop_node>(*loop_state_input));
+  auto si = util::AssertedCast<rvsdg::StructuralInput>(loop_state_input);
+  auto arg = si->arguments.begin().ptr();
+  auto user = get_mem_state_user(arg);
+  auto mux = TryGetOwnerOp<mux_op>(*user);
+  JLM_ASSERT(mux && mux->loop);
+  auto mux_node = rvsdg::TryGetOwnerNode<rvsdg::SimpleNode>(*user);
+  auto lcb_out = loop_constant_buffer_op::create(
+      *mux_node->input(0)->origin(),
+      *mux_node->input(1)->origin())[0];
+  mux_node->output(0)->divert_users(lcb_out);
+  JLM_ASSERT(mux_node->IsDead());
+  remove(mux_node);
+}
+
+void
+decouple_mem_state(rvsdg::Region * region)
+{
+  auto lambda = util::AssertedCast<const jlm::rvsdg::LambdaNode>(region->Nodes().begin().ptr());
+  auto state_arg = GetMemoryStateArgument(*lambda);
```

Comment:
Let's do an assert that the region really only contains a single lambda

Reply:
Added